### PR TITLE
Create Job_scheduling.py

### PR DESCRIPTION
**Pull request by Mohitkumar6122**
### **Describe your change:**
Added a greedy algorithim for job scheduling . this algo is frequently asked in coding interview and it is one of the famous greedy algorithim.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by Mohitkumar6122** [comment]
How to remove trailing whitespaces warning in PEP8 format ?

**Comment by Mohitkumar6122** [comment]
Finaly after 6 builds it passed xD

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/tree/master/scheduling is the place for scheduling algorithms.

**Comment by Mohitkumar6122** [comment]
> https://github.com/TheAlgorithms/Python/tree/master/scheduling is the place for scheduling algorithms.

ohh i just pasted it into greedy i will move it to scheduling ASAP.

**Comment by Mohitkumar6122** [comment]
could you review my PR @cclauss

File: scheduling/job_scheduling.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python
```
Comment (cclauss) [inline scheduling/job_scheduling.py:1]:
```suggestion
#!/usr/bin/env python3
```

File: scheduling/job_scheduling.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+# -*- coding: utf-8 -*-
```
Comment (cclauss) [inline scheduling/job_scheduling.py:2]:
```suggestion
```
This line was only relevant on Python 2

File: scheduling/job_scheduling.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+class Scheduling:
+
+    def __init__(self, jobs):
+        self.jobs = jobs
+
+    def schedule(self, n, deadline):
+        self.sdl = deadline
+        self.J = []
+        self.J.append(self.jobs[1])
```
Comment (cclauss) [inline scheduling/job_scheduling.py:17]:
```suggestion
        self.J = [self.jobs[1]]
```

File: scheduling/job_scheduling.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        self.x = 2
+        while self.x < n:
+            self.K = self.J.copy()
+            self.K.append(self.jobs[self.x])
+            self.x = self.x + 1
+            self.feasibility = self.feasible(self.K, self.sdl)
+            if self.feasibility is True:
```
Comment (cclauss) [inline scheduling/job_scheduling.py:24]:
```suggestion
            if self.feasible(self.K, self.sdl):
```

File: scheduling/job_scheduling.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                self.J = self.K.copy()
+
+        return self.J
+
+    def feasible(self, K, fdl):
+        self.tmp = K
+        self.isFeasible = True
```
Comment (cclauss) [inline scheduling/job_scheduling.py:31]:
```suggestion
        self.is_feasible = True
```

File: scheduling/job_scheduling.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        self.j = 1
+        self.k = 0
+
+        while self.i < len(self.tmp):
+            while self.j < len(self.tmp):
+                self.index1 = self.jobs.index(self.tmp[self.i])
+                self.index2 = self.jobs.index(self.tmp[self.j])
+                self.j = self.j + 1
+                if fdl[self.index1] > fdl[self.index2]:
+                    (self.tmp[self.i], self.tmp[self.j]) = \
```
Comment (cclauss) [inline scheduling/job_scheduling.py:43]: Please format this code with psf/black to remove the backslash.

File: scheduling/job_scheduling.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                        (self.tmp[self.j], self.tmp[self.i])
+            self.i = self.i + 1
```
Comment (cclauss) [inline scheduling/job_scheduling.py:45]:
```suggestion
            self.i += 1
```

File: scheduling/job_scheduling.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        return self.J
+
+    def feasible(self, K, fdl):
+        self.tmp = K
+        self.isFeasible = True
+
+        self.i = 0
+        self.j = 1
+        self.k = 0
```
Comment (cclauss) [inline scheduling/job_scheduling.py:35]:
```suggestion
        i = 0
        j = 1
        k = 0
```
Do not bind these variables to the object.

File: scheduling/job_scheduling.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                    (self.tmp[self.i], self.tmp[self.j]) = \
+                        (self.tmp[self.j], self.tmp[self.i])
+            self.i = self.i + 1
+
+        while self.k < len(self.tmp):
+            self.job = self.tmp[self.k]
+            self.jobindex = self.jobs.index(self.job)
+            self.dlineval = fdl[self.jobindex]
+            self.ftest = self.k + 1
+            self.k = self.k + 1
```
Comment (cclauss) [inline scheduling/job_scheduling.py:52]:
```suggestion
            self.k += 1
```

File: scheduling/job_scheduling.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    n = 7  # Number of jobs
+    jobs = [  # jobs id
+        0,
+        1,
+        2,
+        3,
+        4,
+        5,
+        6
+    ]
```
Comment (cclauss) [inline scheduling/job_scheduling.py:70]:
```suggestion
    jobs = [0, 1, 2, 3, 4, 5, 6]  # jobs id
```

Comment (cclauss) [reply scheduling/job_scheduling.py:70]: Same with the lines below.

File: scheduling/job_scheduling.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    sins = Scheduling(jobs)
+    sjobs = sins.schedule(n, deadline)
+    print('\n Jobs', sjobs)
+
+    finalprofit = []
+    finaldl = []
+
+    for c in range(len(sjobs)):
+        item = sjobs[c]
```
Comment (cclauss) [inline scheduling/job_scheduling.py:102]:
```suggestion
    for i, item in enumerate(sjobs):
```
Use i for integers and c for characters.

File: scheduling/job_scheduling.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        jobsindex = jobs.index(item)
+        finalprofit.append(profit[jobsindex])
+        finaldl.append(deadline[jobsindex])
+
+    print('\n profit', finalprofit)
+    print('\n Deadline', finaldl)
```
Comment (cclauss) [inline scheduling/job_scheduling.py:108]:
```suggestion
    print(f'\n profit {finalprofit}')
    print(f'\n Deadline {finaldl}')
```

**Comment by cclauss** [comment]
Try to bind as few things to the object as you can.  `x` will always be faster and easier to understand (and debug) than `my_object.x`.

File: scheduling/job_scheduling.py (hunk 6 above)
Comment (Mohitkumar6122) [reply scheduling/job_scheduling.py]: where can i find psf/black formatter  ??

**Comment by Mohitkumar6122** [comment]
> Try to bind as few things to your objects as you can. `x` will always be faster and easier to understand (and debug) than `my_object.x`.

thanks for the suggestion , i will keep it in mind.

File: scheduling/job_scheduling.py [hunk 13]
```diff
@@ -1,112 +1,85 @@
...
-        46,
-        52,
-        30,
-        36,
-        56,
-        40
-    ]
+    jobs = [0, 1, 2, 3, 4, 5, 6]  # jobs id
+    deadline = [0, 2, 4, 3, 3, 2, 1]  # deadline to job[i]
+    profit = [0, 46, 52, 30, 36, 56, 40]  # profit associated wrt job[i]
```
Comment (cclauss) [inline scheduling/job_scheduling.py:64]: Organize job data using a namedtuple...
```python
from collections import namedtuple
job = namedtuple("job", "job_id deadline profit")
jobs = [job(0, 0, 0), job(1, 2, 46) ... ]
```

Comment (Mohitkumar6122) [reply scheduling/job_scheduling.py:64]: Actually I am not familiar with namedtuple and how to use it so i need  a bit more time to apply that

File: scheduling/job_scheduling.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+"""
+Author : Mohit Kumar
+Job Sequencing Problem implemented in python
+"""
+from collections import namedtuple
+
+
+class Scheduling:
+    def __init__(self, jobs):
```
Comment (cclauss) [inline scheduling/job_scheduling.py:11]: Python type hints?  doctests?  See CONTRIBUTING.md.

File: scheduling/job_scheduling.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        self.jobs = jobs
+
+    def schedule(self, n, deadline):
```
Comment (cclauss) [inline scheduling/job_scheduling.py:14]: Python type hints?  doctests?  Avoid single-letter variable names especially for function parameters.  See CONTRIBUTING.md.

File: scheduling/job_scheduling.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        self.sdl = deadline
+        self.J = [self.jobs[1]]
```
Comment (cclauss) [inline scheduling/job_scheduling.py:16]:
```suggestion
        self.j = [self.jobs[1]]
```
In Python, uppercase variable names are reserved for constants (do not change during the runtime of the app).

File: scheduling/job_scheduling.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+        while self.x < n:
+            self.K = self.J.copy()
+            self.K.append(self.jobs[self.x])
+            self.x = self.x + 1
+            if self.feasible(self.K, self.sdl):
+                self.J = self.K.copy()
+
+        return self.J
+
+    def feasible(self, K, fdl):
```
Comment (cclauss) [inline scheduling/job_scheduling.py:27]: Python type hints? doctests? Avoid single letter variable names especially on for function parameters. See CONTRIBUTING.md.

Why do we need a fiddle?   Why force the reader to guess at what fdl is?

File: scheduling/job_scheduling.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+            k += 1
+            if self.dlineval < self.ftest:
+                self.is_feasible = False
+                break
+
+        return self.is_feasible
+
+
+def main():
+    n = 7  # Number of jobs
```
Comment (cclauss) [inline scheduling/job_scheduling.py:61]: Why create `n` when you can use `len(jobs)` instead?

File: scheduling/job_scheduling.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+    for i in range(n):
+        cur_job = []
+        cur_job.extend((jobs[i].deadline, jobs[i].profit, jobs[i].jobs))
+        midresult.append(cur_job)
+    midresult.sort(key=lambda k: (k[0], -k[1]))
+    (deadline, profit, jobs) = map(list, zip(*midresult))
+
+    sins = Scheduling(jobs)
+    sjobs = sins.schedule(n, deadline)
+    print("\n Jobs", sjobs)
```
Comment (cclauss) [inline scheduling/job_scheduling.py:83]: f-string

File: scheduling/job_scheduling.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        job(2, 4, 52),
+        job(3, 3, 30),
+        job(4, 3, 36),
+        job(5, 2, 56),
+        job(6, 1, 40),
+    ]
+
+    midresult = []
+    for i in range(n):
+        cur_job = []
```
Comment (cclauss) [inline scheduling/job_scheduling.py:75]:
```suggestion
        current_job = []
```
We can afford the four extra characters to make reading easier.

File: scheduling/job_scheduling.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        cur_job.extend((jobs[i].deadline, jobs[i].profit, jobs[i].jobs))
+        midresult.append(cur_job)
```
Comment (cclauss) [inline scheduling/job_scheduling.py:77]: What is a midresult and why does the reader need to guess?

**Review by cclauss — CHANGES_REQUESTED** [review]

File: scheduling/job_scheduling.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+    midresult.sort(key=lambda k: (k[0], -k[1]))
+    (deadline, profit, jobs) = map(list, zip(*midresult))
+
+    sins = Scheduling(jobs)
```
Comment (cclauss) [inline scheduling/job_scheduling.py:81]: sins is a poor variable name.  `scheduled_jobs` might be more self-documenting.

File: scheduling/job_scheduling.py [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+        jobsindex = jobs.index(item)
+        finalprofit.append(profit[jobsindex])
+        finaldl.append(deadline[jobsindex])
+
+    print(f"\n Profit {finalprofit}")
+    print(f"\n Deadline {finaldl}")
+
+
+if __name__ == "__main__":
+    main()
```
Comment (Abdujabbor) [inline scheduling/job_scheduling.py:134]: as I know for lunching tests we need add these lines

`import doctest

doctest.testmod()` 

am I right @cclauss ? if I am wrong, sorry for that. :)

Comment (cclauss) [reply scheduling/job_scheduling.py:134]: Travis CI knows how to find the doctest and is running them
https://travis-ci.com/github/TheAlgorithms/Python/builds/186063672#L779

It is optional to add `doctest.testmod()` to main() which will cause the tests to run locally for users.

**Comment by Mohitkumar6122** [comment]
@cclauss could you review my PR ?

**Comment by cclauss** [comment]
###  Create a Job class

I hope these changes help to clarify...

I have rearranged the code in a way that will BREAK THE TESTS.   I make this clearer, we will have a Job class and will remove the Schedule class and replace it with functions.  Let's keep the Job class super simple.  Its only goal is to carry around the deadline and profit for each Job so we do not move around separate list of ints.

With these changes, each of our three main functions takes in a list of jobs.  We should keep things flexible so that we could create `test_jobs_two` and run it thru all functions and get different results.

**Comment by Mohitkumar6122** [comment]
> ### Create a Job class
> I hope these changes help to clarify...
> 
> I have rearranged the code in a way that will BREAK THE TESTS. I make this clearer, we will have a `Job` class and will remove the `Schedule` class and replace it with functions. Let's keep the `Job` class super simple. Its only goal is to carry around the `deadline` and `profit` for each `Job` so that we no longer move around separate lists of ints.
> 
> With these changes, each of our three main functions takes in a list of jobs. We should keep things flexible so that we could create `test_jobs_two` and run it thru all functions and get different results.

your changes made code super easy and understandable but problem is that it isn't running any more :worried:

**Comment by cclauss** [comment]
> problem is that it isn't running anymore

Yes...  I do not understand the algorithm so you will have to make it work again with the Jobs class and the function signatures that I provided.

**Comment by Mohitkumar6122** [comment]
> > problem is that it isn't running anymore
> 
> Yes... I do not understand the algorithm so you will have to make it work again with the Jobs class and the function signatures that I provided.

sure i will work on it ASAP but my exams are coming so maybe i will start working on it after 2 weeks :smile:

**Comment by spamegg1** [comment]
> 1     E302 expected 2 blank lines, found 1
1     E305 expected 2 blank lines after class or function definition, found 1
54    F821 undefined name 'self'

**Comment by cclauss** [comment]
Yes @spamegg1 that was already discussed in the comments above.

**Pull request closed without merging**